src/Infrastructure/Repository/TaskYamlRepository.php

```diff
@@ -87,52 +109,99 @@ public function __construct(
     {
-        $taskDirSetting = $this->settingRepository->findOneByPath('extension_dirs');
-
-        if (!$taskDirSetting || !is_array($taskDirSetting->getValues())) {
-            throw new MissingSettingException('extension_dirs are not configured properly');
+        if (!$this->taskListData || !$this->taskSetsData) {
+            $this->readTaskYaml();
         }
 
-        $tasks = [];
-        $taskListData = [];
-        $taskSetsData = [];
-
-        $finder = $this->fileFinder
-            ->in($this->findExistedDirectories($taskDirSetting->getValues()))
-            ->name('*.yaml');
-
-        //read task from path, parse and create Task, later use DB for querying
-        foreach ($finder->files() as $taskFile) {
-            $taskData = $this->yamlParser->parse($taskFile->getContents());
-
-            if ($taskData['type'] === static::TASK_SET_TYPE) {
-                $taskSetsData[$taskData['id']] = $taskData;
-            } else {
-                $taskListData[$taskData['id']] = $taskData;
-            }
-        }
+        $this->taskListData = $this->manifestValidation->validateAndNormalize(TaskManifestConfigTreeBuilderFactory::NAME, $this->taskListData);
+        $this->taskSetsData = $this->manifestValidation->validateAndNormalize(TaskSetManifestConfigTreeBuilderFactory::NAME, $this->taskSetsData);
 
-        foreach ($taskListData as $taskData) {
-            $task = $this->buildTask($taskData, $taskListData, $tags);
+        $tasks = [];
+        foreach ($this->taskListData as $taskData) {
+            $task = $this->buildTask($taskData, $this->taskListData, $tags);
             $tasks[$task->getId()] = $task;
         }
 
         $existingTasks = array_merge($this->existingTasks, $tasks);
 
-        foreach ($taskSetsData as $taskData) {
-            $task = $this->buildTaskSet($taskData, $taskListData, $existingTasks, $tags);
+        foreach ($this->taskSetsData as $taskData) {
+            $task = $this->buildTaskSet($taskData, $this->taskListData, $existingTasks, $tags);
             $tasks[$task->getId()] = $task;
         }
 
         return array_merge($tasks, $this->existingTasks);
     }
 
+    /**
+     * Specific method for task validation
+     *
+     * @param string $taskId
+     * @param bool $includeTaskSet
+     *
+     * @return bool
+     */
+    public function isTaskIdExist(string $taskId, bool $includeTaskSet = true): bool
+    {
+        $this->readTaskYaml();
+
+        return isset($this->taskListData[$taskId]) ?:
+            isset($this->existingTasks[$taskId]) ?:
+            isset($this->taskSetsData[$taskId]) && $includeTaskSet ?:
+            false;
+    }
+
+    /**
+     * Specific method for task set validation
+     *
+     * @param array<string> $taskIds
+     * @param bool $includeTaskSet
+     *
+     * @return array<string, array<string>>
+     */
+    public function returnTaskPlaceholders(array $taskIds, bool $includeTaskSet = false): array
```

Comment:
Separate this method into three to handle different sources of placeholders

Reply:
it's not a different source. this logic should be only in YAML repository. That's why I wrote specifications for 2 methods. When we are going to validate one task we must have access to others.  the problem is that we can't map yaml to object before validation because the structure can be wrong. we still have not refactored this logic.

Reply:
Will be changed after  in the separate task yaml repository refactoring task.